crates/frontend/src/contextmenu/mod.rs

```diff
@@ -1,17 +1,179 @@
 use std::rc::Rc;
 use futures_signals::signal::Mutable;
+use dominator::{Dom, html, clone, events};
+use wasm_bindgen::JsValue;
+use crate::{vfs::{Directory, File}, DEFAULT_DIRECTORY_MODE, DEFAULT_FILE_MODE};
 
-pub struct ContextMenuState {
-    pub show_menu: Mutable<bool>,
-    pub menu_position: Mutable<(i32, i32)>,
+#[derive(Clone, Copy)]
+pub enum Menutype {
+    File, 
+    Directory
+}
+
+pub struct ContextMenu {
+    // visibility and position of contextmenu
+    pub show: Mutable<bool>,
+    pub position: Mutable<(i32, i32)>,
+    // to split the contextmenu separately for file and folder
+    pub menu_type: Mutable<Option<Menutype>>,
+    // to see which file or folder is clicked
+    pub target_object: Mutable<Option<Rc<Directory>>>
 }
 
 
-impl ContextMenuState {
+impl ContextMenu {
     pub fn new() -> Rc<Self> {
         Rc::new(Self {
-            show_menu: Mutable::new(false),
-            menu_position: Mutable::new((0, 0)),
+            show: Mutable::new(false),
+            position: Mutable::new((0, 0)),
+            menu_type: Mutable::new(None),
```

Comment:
I would consider having a `menu_type` as an `Option<MenuType>` [code smell](https://en.wikipedia.org/wiki/Code_smell) in Rust. The problem is that it is possible to create an instance of `ContextMenu` without a "type" which feels like an invalidate state. I would remove the `Option` from `menu_type`.

Reply:
I have made an `Enum` already for File and Folder, should I get rid of the `menu_type` variable all together and just integrate the `MenuType` `Enum` in the `target_object` variable?

Reply:
Ideally the state of the program would reflect the state of the UI. If there is no context menu in the UI, then an instance of the context menu should not exist in the program's memory either.

Reply:
I made it a `Mutable<Option<ContextMenu>>` so that we can use `child_signal` to react is there is `Some` context menu or `None`.